### DISCUSSION TITLE
Request to Inverter over CAN to maintain good charging scheme for battery

### DISCUSSION THREAD

**Discussion by paulsteigel** [ArminJo/JK-BMSToPylontechCAN#18] [General]
Hi Armin,
I have successfully update the Battery Capacity over your work with small modification.
`struct PylontechCANSpecificationsFrameStruct {
    struct PylontechCANFrameInfoStruct PylontechCANFrameInfo = { PYLON_CAN_BATTERY_SPECIFICATIONS_FRAME_ID, 2 }; // 0x379
    struct {
        uint16_t CapacityAmpereHourLowByte;
        uint16_t CapacityAmpereHourHighByte;
    } FrameData;
    void fillFrame(struct JKReplyStruct *aJKFAllReply) {
        (void) aJKFAllReply; // To avoid [-Wunused-parameter] warning
        FrameData.CapacityAmpereHourLowByte = lowByte(JKComputedData.TotalCapacityAmpereHour);
        FrameData.CapacityAmpereHourHighByte = highByte(JKComputedData.TotalCapacityAmpereHour);
    }
};`

However, there are improvement that would be great if your project can have such as:
+ Proactive Charging request over CAN (base on calculating voltage, SOC) so the inverter can apply a smooth, cooldown charging scheme to the battery. This would be good for battery life; For this approach, I came by Uksa007 and found his approach of interesting and valuable for porting to your project (see here https://github.com/Uksa007/esphome-jk-bms-can/blob/main/esp32-example-can.yaml#L17 with idea of Absorption voltage and code from line 186). The approach is quite simple:
>> base on the voltage, soc, time... calculate the time for high wattage charging; then When Voltage, SOC reaching 80%, send Inverter a new value of charging current, wattage >> we finally gain a linear/cooling-down charging curve for the battery.
>> All of those will be done via value of current over CAN bus (JKBMS send the default value but Arduino can calculate and send Inverter a new value to manipulate the charging process)
+ Connecting multiple JK-BMS (Battery) in parallels for adding up capacity: If we can do this over modbus, the contribution of this project shall be of great value to users like all of us.
I would like to raise this point of discussion for your consideration, if possible, please be so kind to help us. With your coding, I am gaining a lot.
Thank you very much for your efforts!
Best,
Ngoc

**Comment by ArminJo** [comment]
> * Connecting multiple JK-BMS (Battery) in parallels for adding up capacity:
What data must be merged?
Only the flags and the currents?

**Reply by paulsteigel** [reply]
Hi Armin,
Thank you for  your question! You are right that, adding more battery packs in parallel will help increasing:
1. Total capacity (Ah)
2. Current (A);
3. Number of module: Inverter will see how many pack are adding up
However, few things shall have to be considered:
1. Warning on cells status (if any);
2. Online/offline status of pack: needed for calculating number of pack those are currently feeding the inverter.
I am thinking of using 2 additional RJ45 jacks for connecting slave battery in parallels using mod-bus with Dip switch to set address of each battery.
Only the master battery will use CAN to interface with the inverter. Other battery will talk to the master battery only over the two additional rj45 jacks.
That's should be too much already. I hope you will be able to have some time to help the communities on these!
Best
Ngoc

**Reply by paulsteigel** [reply]
Hi Armin,
I have applied the charging scheme with gradual reduction of charging voltage from full voltage (JKComputedData.BatteryFullVoltage10Millivolt/10) to float voltage (JKComputedData.BatteryVoltage10Millivolt/10) when SOC is at 80% or above. I went through your code to check the for the FloatVoltage and found that your Float voltage is calculated  in JK-BMS.hpp (void fillJKComputedData).
Normally, in this charging state, voltage should be gradually down from full voltage (3.65*number of cell - applied with LiFePO4) to Float voltage (3.2*number of Cell). 
However, I realised your float voltage is calculated as below (this figure is vary from time to time and can not be used as the lower bound for applying charging voltage):
**JKComputedData.BatteryVoltageFloat = JKComputedData.BatteryVoltage10Millivolt;**
I tried checking your code to find the Float voltage but dont see anywhere with this. It attached to updated battery voltage, so upon changing, the battery charge will not reduced in voltages but increasing to the full voltage established in JKBMS (Here my setting is 3.4V for OVP).
Can you advise me on where I can obtain correct Float voltage for a cell or the entire battery. Or can you be so kind to modify charging with this reading (https://sunonbattery.com/properly-charge-lifepo4-battery/)
Best,
PS: I also found out that Luxpower will use 0x373 (cell infor) to accept MaxCellVoltage and MinCellVoltage, but not yet figure out battery cyclecount and max/min cell temperature (See image)
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/fbc5e393-fb20-4255-bff7-433eed59eab3)

Charging curve with voltage control: It seems to be ok no sudden increasing of charging power as before, also no change to current limit shown on Luxpower reading. Only minor improving on a fixed floating voltage is needed!
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/44d8036f-856b-4a7d-8de8-26b1c100cd57)

**Comment by ArminJo** [comment]
???
what is the problem with the float value?

**Reply by paulsteigel** [reply]
Hi Armin, 
It increased over time as it stick to battery voltage (following your code). So it will not have effect on reducing the charging voltage to the floating level as the SOC goes up!
I did not enable the: **#define CAN_DATA_MODIFICATION**  for can modification but control the charging voltage directly in the filling frame for currentVoltageLimiting frame!
Best,
Ngoc

**Reply by ArminJo** [reply]
> It increased over time as it stick to battery voltage 

This is the definition 😀 of it.
Are you looking for JKComputedData.BatteryFullVoltage10Millivolt, sent in CAN 0x351?

**Reply by paulsteigel** [reply]
Hi Armin,
Thanks for your information, I finally found a way to deal with it:
1. Peak voltage to start reducing (at 80% soc) = 3.65 V for LFP and 4.3 V for lithium or use the voltage that the inverter applied when hitting the SOC of 80%.
2. Float Voltage to end the charge = OVP voltage set in the BMS.

With SOC increasing, the charging voltage will reduce accordingly to the last value expected. 
I observe most of users will set OVP at 3.4 for 100% and 3.2 or bit less at 0% on BMS to protect over discharging/charging.
With this, the charging will be smooth enough!
Great thanks to your helps and advices,
Ngoc.
